Framework/Core/src/DataSpecUtils.cxx

```diff
@@ -233,26 +263,160 @@ ConcreteDataMatcher DataSpecUtils::asConcreteDataMatcher(OutputSpec const& spec)
+                      },
+                      [](DataDescriptorMatcher const& matcher) {
+                        return extractDataType(matcher);
+                      } },
+                    spec.matcher);
+}
+
+DataDescriptorMatcher DataSpecUtils::dataDescriptorMatcherFrom(ConcreteDataTypeMatcher const& dataType)
+{
+  auto timeDescriptionMatcher = std::make_unique<DataDescriptorMatcher>(
+    DataDescriptorMatcher::Op::And,
+    DescriptionValueMatcher{ dataType.description.as<std::string>() },
+    StartTimeValueMatcher(ContextRef{ 0 }));
+  return std::move(DataDescriptorMatcher(
+    DataDescriptorMatcher::Op::And,
+    OriginValueMatcher{ dataType.origin.as<std::string>() },
+    std::move(timeDescriptionMatcher)));
+}
+
+InputSpec DataSpecUtils::matchingInput(OutputSpec const& spec)
+{
+  return std::visit(overloaded{
+                      [&spec](ConcreteDataMatcher const& concrete) -> InputSpec {
+                        return InputSpec{
+                          spec.binding.value,
+                          concrete.origin,
+                          concrete.description,
+                          concrete.subSpec,
+                          spec.lifetime
+                        };
+                      },
+                      [&spec](ConcreteDataTypeMatcher const& dataType) -> InputSpec {
+                        auto&& matcher = DataSpecUtils::dataDescriptorMatcherFrom(dataType);
+                        return InputSpec{
+                          spec.binding.value,
+                          std::move(matcher)
+                        };
+                      } },
+                    spec.matcher);
 }
 
 std::optional<header::DataHeader::SubSpecificationType> DataSpecUtils::getOptionalSubSpec(OutputSpec const& spec)
 {
-  ConcreteDataMatcher concrete = DataSpecUtils::asConcreteDataMatcher(spec);
-  return concrete.subSpec;
+  return std::visit(overloaded{
+                      [](ConcreteDataMatcher const& concrete) -> std::optional<header::DataHeader::SubSpecificationType> {
+                        return std::make_optional(concrete.subSpec);
+                      },
+                      [](ConcreteDataTypeMatcher const&) -> std::optional<header::DataHeader::SubSpecificationType> {
+                        return {};
+                      } },
+                    spec.matcher);
+}
+
+std::optional<header::DataHeader::SubSpecificationType> DataSpecUtils::getOptionalSubSpec(InputSpec const& spec)
+{
+  // FIXME: try to address at least a few cases.
+  return std::visit(overloaded{
+                      [](ConcreteDataMatcher const& concrete) -> std::optional<header::DataHeader::SubSpecificationType> {
+                        return std::make_optional(concrete.subSpec);
+                      },
+                      [](DataDescriptorMatcher const&) -> std::optional<header::DataHeader::SubSpecificationType> {
+                        return {};
```

Comment:
If the inputSpec is created out of a query, I can see that it always becomes a DataDescriptorMatcher, even if it actually has a SubSpec specified. So the assumption that it never has a SubSpec is a bit wrong, could you actually check for it here?

Reply:
Yes, you are right. I will update the code to extract the subSpec from the query.